starfish/imagestack/parser/crop.py

```diff
@@ -0,0 +1,198 @@
+        permitted_chs : Optional[Collection[int]]
+            The channels in the original dataset to load into the ImageStack.  If this is not set,
+            then all channels are loaded into the ImageStack.
+        permitted_zlayers : Optional[Collection[int]]
+            The z-layers in the original dataset to load into the ImageStack.  If this is not set,
+            then all z-layers are loaded into the ImageStack.
+        x_slice : Optional[Union[int, slice]]
+            The x-range in the x-y tile that is loaded into the ImageStack.  If this is not set,
+            then the entire x-y tile is loaded into the ImageStack.
+        y_slice : Optional[Union[int, slice]]
+            The y-range in the x-y tile that is loaded into the ImageStack.  If this is not set,
+            then the entire x-y tile is loaded into the ImageStack.
+        """
+        self._permitted_rounds = set(permitted_rounds) if permitted_rounds else None
+        self._permitted_chs = set(permitted_chs) if permitted_chs else None
+        self._permitted_zlayers = set(permitted_zlayers) if permitted_zlayers else None
+        self._x_slice = x_slice
+        self._y_slice = y_slice
+
+    def filter_tilekeys(self, tilekeys: Collection[TileKey]) -> Collection[TileKey]:
+        """
+        Filters tilekeys for those that should be included in the resulting ImageStack.
+        """
+        results: MutableSequence[TileKey] = list()
+        for tilekey in tilekeys:
+            if self._permitted_rounds is not None and tilekey.round not in self._permitted_rounds:
+                continue
+            if self._permitted_chs is not None and tilekey.ch not in self._permitted_chs:
+                continue
+            if self._permitted_zlayers is not None and tilekey.z not in self._permitted_zlayers:
+                continue
+
+            results.append(tilekey)
+
+        return results
+
+    @staticmethod
+    def _crop_axis(size: int, crop: Optional[Union[int, slice]]) -> Tuple[int, int]:
+        """
+        Given the size of along an axis, and an optional cropping, return the start index
+        (inclusive) and end index (exclusive) of the crop.  If no crop is specified, then the
+        original size (0, size) is returned.
+        """
+        # convert int crops to a slice operation.
+        if isinstance(crop, int):
+            crop = slice(crop, crop + 1)
+
+        # convert start and stop to absolute values.
+        start: int
+        if crop is None or crop.start is None:
+            start = 0
+        elif crop.start is not None and crop.start < 0:
+            start = max(0, size + crop.start)
+        else:
+            start = min(size, crop.start)
+
+        stop: int
+        if crop is None or crop.stop is None:
+            stop = size
+        elif crop.stop is not None and crop.stop < 0:
+            stop = max(0, size + crop.stop)
+        else:
+            stop = min(size, crop.stop)
```

Comment:
If someone provides crop parameters outside the range of the axis should we through an error instead of setting the start and stop to 0 and the size?

Reply:
I don't have a strong opinion here.  I was obeying the semantics of python slicing, but I'm happy to change it up if you think it's better, @shanaxel42.